felix/dataplane/linux/qos_controls.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intdataplane
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/projectcalico/calico/felix/dataplane/linux/bwp"
+	"github.com/projectcalico/calico/felix/proto"
+)
+
+func (m *endpointManager) maybeUpdateQoSBandwidth(old, new *proto.WorkloadEndpoint) error {
+	var errs []error
+	var oldIngressBw, oldEgressBw, newIngressBw, newEgressBw int64
+	var oldIngressBurst, oldEgressBurst, newIngressBurst, newEgressBurst int64
+	if old != nil && old.QosControls != nil {
+		oldIngressBw = old.QosControls.IngressBandwidth
+		oldEgressBw = old.QosControls.EgressBandwidth
+		oldIngressBurst = old.QosControls.IngressBurst
+		oldEgressBurst = old.QosControls.EgressBurst
+	}
+
+	if new != nil && new.QosControls != nil {
+		newIngressBw = new.QosControls.IngressBandwidth
+		newEgressBw = new.QosControls.EgressBandwidth
+		newIngressBurst = new.QosControls.IngressBurst
+		newEgressBurst = new.QosControls.EgressBurst
+	}
+
+	if oldIngressBw != newIngressBw || oldIngressBurst != newIngressBurst {
+		if oldIngressBw != 0 {
+			err := removeIngressQdisc(old.Name)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error removing ingress qdisc: %v", err))
+			}
+		}
+		if newIngressBw != 0 {
+			err := addIngressQdisc(newIngressBw, newIngressBurst, new.Name)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error adding ingress qdisc: %v", err))
+			}
+		}
+	}
+
+	if oldEgressBw != newEgressBw || oldEgressBurst != newEgressBurst {
+		if oldEgressBw != 0 {
+			err := removeEgressQdisc(old.Name)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error removing egress qdisc: %v", err))
+			}
+		}
+		if newEgressBw != 0 {
+			err := addEgressQdisc(newEgressBw, newEgressBurst, new.Name)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error adding egress qdisc: %v", err))
+			}
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+func addIngressQdisc(bw, burst int64, intf string) error {
+	return bwp.CreateIngressQdisc(uint64(bw), uint64(burst), intf)
+}
```

Comment:
Not really about this code specifically, but... what does it actually mean to impose an _ingress_ bw limit?

With _egress_ I think I understand: it's about the queuing of outbound packets for transmission, and if the buffer fills up, there's back pressure on the sending application.

But with _ingress_ there's no local application to apply back pressure to, so how does that work?  Do excess inbound packets just get dropped?

Reply:
Yes, the packets are dropped (as far as I understand it), and I guess it would in some convoluted way result in the sender throttling its transmissions (assuming TCP)? This was implemented for feature parity (and "forwards compatibility") with the cni bandwidth plugin, which does both ingress and egress limiting